### DISCUSSION TITLE
Workflow for active development of a library (dependency) and application

### DISCUSSION THREAD

**Discussion by 0xjairo** [microsoft/vcpkg#20468] [Q&A] [answered]
Is there a recommended workflow for working with libraries (dependencies) that are currently under development?  We have an internal vcpkg registry with custom libraries that are under development. In addition, the application that uses these libraries is also under development.

When we add a new feature to a dependency, and we want the application developers to test this new feature they either have to 

1. Clone the registry repo, clone the dependency repo, modify the port in the cloned registry to point `SOURCE_PATH` to the cloned dependency, and re-build configure CMake with` -DVCPKG_OVERLAY_PORTS="<path-to-modified-port>...`. Alternatively we can...
2. Release a development version into the registry and incur the overhead of releasing to the vpckg-registry (update the REF and/or port version in the portfile, possibly update vcpkg.json too, update the versions/ database, and update the `baseline.json`). This is cumbersome if the application developers find issues that require changes to the library often.

Curious what others are doing in this type of situation of there's a better workflow.

Thanks!

**Comment by thxkiwi** [comment]
I have a similar scenario but I don't think it's exactly yours (https://github.com/microsoft/vcpkg/discussions/19784)

See the short discussion on my post. I created a script that removes installed and cached artifacts and populates them with the artifacts most recently built. The library and the application using the lib are in the same build tree so HEAD is the same point for both and can never the HEAD for application will always build against the same HEAD for the lib.

I considered both of the possible solutions you've mentioned and decided neither were ideal because both are not in the natural workflow for development. Adding them to the workflow makes it cumbersome and error prone.

**Comment by 0xjairo** [comment, marked as answer]
One thing I've tried today was to utilize CMake's [`<package_name>_ROOT`](https://cmake.org/cmake/help/v3.21/variable/PackageName_ROOT.html#variable:%3CPackageName%3E_ROOT) variable to use the dependency from an installation directory separate from vcpkg.

What I do is in the library's source repository:  `cmake -B build -S . -G Ninja -DCMAKE_INSTALL_PREFIX=<install_folder>`, where `<install_folder>` is an arbitrary directory on my machine (Note: windows requires putting quotes around the entire variable, see footnote[1] below). Then in the app's directory I configure CMake with `cmake -B build -S . -G Ninja -Dmy_cool_library_ROOT=<install_folder>`.   This supersedes the installation prefix used by vcpkg for `my_cool_library` **only** with the `_ROOT` path given, the rest of vcpkg dependencies resolve normally.

[1]: In windows do put quotes around the variable and the path like this: `-D "CMAKE_INSTALL_PREFIX=../../install_cm"`
